Month 8 / step 1

Shiker2032 opened the pull request.

Shiker2032 committed:
Добавление ингредиентов в конструктор

Shiker2032 committed:
Сделал проверку дубликата булки

Shiker2032 committed:
Сделал расчет стоимости

Shiker2032 committed:
Сделал отображение номера заказа в попапе

Shiker2032 committed:
Сделал стоимость через редюсер

Shiker2032 committed:
начал рефакторить

gennady-bars commented on src/components/App/App.jsx:
```diff
@@ -41,29 +41,59 @@ function App() {
-  function getState() {
+  const handleOrderClick = () => {
+    const orderInfo = order.map((ingredients) => ingredients._id);
+    fetch("https://norma.nomoreparties.space/api/orders", {
```
`'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `baseUrl` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было сменить адрес, но не менять эндпоинты.

gennady-bars commented on src/components/App/App.jsx:
```diff
@@ -41,29 +41,59 @@ function App() {
+        ingredients: orderInfo,
+      }),
+    })
+      .then(parseResponse)
```
Хорошо проверяете ответы от сервера

gennady-bars commented on src/components/App/App.jsx:
```diff
@@ -41,29 +41,59 @@ function App() {
+          />
+          <BurgerConstructor
+            ingredients={ingredients}
+            order={order}
```
`            order={order}
`

Вы сделали контекст `OrderContext` специально, чтобы `order` раздавать, а все равно передаете в пропсы тут. Нужно брать уже из контекста внутри компонентов данные

gennady-bars commented on src/components/burger-constructor/BurgerConstructor.jsx:
```diff
@@ -1,56 +1,91 @@
+    };
+  }
+ 
+  const [state, dispatch] = useReducer(reducer, initialState);
```
Хорошо сделали редьюсер

gennady-bars commented on src/components/burger-constructor/ConstructorItem.jsx:
```diff
@@ -1,11 +1,13 @@
 import { ConstructorElement } from "@ya.praktikum/react-developer-burger-ui-components";
 
-function ConstructorItem({ ingredient, type }) {
+function ConstructorItem({ ingredient, type, children }) {
```
`ConstructorItem` забыли типизировать

gennady-bars commented on src/components/Burger-ingredients/BurgerIngredients.jsx:
```diff
@@ -10,7 +10,6 @@ import {
 } from "@ya.praktikum/react-developer-burger-ui-components";
 
 const BurgerIngredients = ({ ingredients, onClick }) => {
```
в `BurgerIngredients` еще есть `onClick`. Его тоже нужно типизировать

gennady-bars reviewed (commented):
Здравствуйте. (**Нужно развернуть общий комментарий ↓**)

Работа проделана огромная, но есть некоторые недочеты:



* Файл `App.test.js` нужно удалить, так как он не относится к проекту
* `'https://norma.nomoreparties.space/api` везде повторяется в запросах. Нужно вынести в константу `baseUrl` и импортировать везде в коде проекта, добавляя в конце изменяющуюся часть (эндпоит). Это обычная практика, чтобы одним движением можно было сменить адрес, но не менять эндпоинты.
* у Вас должен быть единый обработчик `Escape` внутри компонента `Modal`. Не нужно под каждый попап делать отдельную функцию.  Внутри `Modal` нужно объявить обработчик внутри `useEffect`, тогда ссылка на функцию не будет меняться и только единый обработчик будет
* у ингредиентов в заказе (между булками) должна быть иконка удаления, а не замочек. Скрин http://joxi.ru/J2bzbWltq5WZk2 
* текст `верх/низ` должен быть в конце названия булок. Скрин http://joxi.ru/V2V749KCBqD6EA
* в `BurgerConstructor` Вы больше не используете `ingredients`, поэтому нужно удалить типизацию
* в `BurgerIngredients` еще есть `onClick`. Его тоже нужно типизировать
* `ConstructorItem`, `OrderDetails` забыли типизировать
* Нужно посмотреть все компоненты и типизировать их `пропсы`, если где-то пропустили. 





**Можно лучше**

* По идее нужно было сделать только один контекст `BurgerContext` и обернуть в него всё приложение. В `value` ему передавать массив ингредиентов, полученный с сервера. А уже внутри компонентов `BurgerIngredients` и `BurgerConstructor` брать этот массив и делать с ним то, что нужно.
* лучше не давать нажать на кнопку `Оформить заказ`, пока пользователь не добавил булку в заказ.
* При кликах на табы `Булки, Соусы, Начинки` можно сделать логику `плавного` скролла до секции. Это можно сделать с помощью хука `useRef` и методом ` scrollIntoView`. Нужно установить для 3х секций 3 `useRef` и следить за изменением табов при кликах на них через функцию клика. Как меняется таб, то запускается ` scrollIntoView ` со значением ` behavior: "smooth"`
Подробнее про `useRef` тут https://ru.reactjs.org/docs/hooks-reference.html#useref
Подробнее про ` scrollIntoView` тут https://developer.mozilla.org/ru/docs/Web/API/Element/scrollIntoView


Исправьте, пожалуйста, недочеты и работа будет принята. Пожалуйста, проверьте работоспособность проекта и наличие возможных ошибок в консоли браузера (кнопка `F12`) перед отправкой на ревью. 


Удачного рефакторинга кода.

Shiker2032 committed:
Внес исправления

Удалил ненужное
Вынес url в API
Сделал единую обработку Esc
Сделал иконки удаления
Поправил текст
Типизировал пропсы
Скрыл оформление заказа до выбора булки
Сделал плавный скролл

gennady-bars reviewed (approved):
Поздравляю! Ваша работа принята.

Вы отлично потрудились.

Удачного дальнейшего обучения.


* кое-что нужно удалить из пропсов. Скрин http://joxi.ru/Q2KZl97HOqYaVm . Это больше не используется в компонентах
